### PR TITLE
Feature/jwt

### PR DESCRIPTION
**Pull request by HurleySA**
## Description

Adding JWT to application using jsonwebtoken lib.

## Related Issues
https://trello.com/c/vbe6Oemv/12-cria%C3%A7%C3%A3o-de-jwt-no-backend-para-autentica%C3%A7%C3%A3o

## Progress

Use checkboxes to show that you did and what you have to do. For example:

- [x] New endpoint created;
- [x] New folder to Middlewares
- [ x] Authetication running 

## How to test it

The authetication is only applied on the Put Route: http://localhost:3333/api/users 

To request this endpoint is necessary to put Authorization on Header. Create a new Authorization Token using route: http://localhost:3333/api/auth, passing {username, password}

## Visual reference

Screenshot_16
Screenshot_17
Screenshot_18

Please include screenshots, gifs or recordings

## Change logs

JWT

### CODE REVIEW

File: src/controllers/AuthControler.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
+import { Request, Response } from "express";
+import { StatusCodes } from "http-status-codes";
+import jwt from 'jsonwebtoken';
+import { prismaClient } from "../database/prismaClient";
+
+export class AuthController {
+    async authenticate(request: Request, response: Response){
+        const { username, password} = request.body;
+        const secret = process.env.SECRET!;
```
Comment (viniciuslacerda0) [inline src/controllers/AuthControler.ts]:
```suggestion
        const secret = process.env.SECRET ?? '';
```

Comment (viniciuslacerda0) [reply src/controllers/AuthControler.ts]: you can also make an if to see if exists and if not send a server error

File: src/middlewares/authMiddleware.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+export default function authMiddleware(request: Request,response: Response, next: NextFunction){
+    const { authorization } = request.headers;
+
+    if(!authorization) {
+        return response.sendStatus(StatusCodes.UNAUTHORIZED)
+    }
+
+    const token = authorization.split(' ')[1];
+    const secret = process.env.SECRET!
```
Comment (viniciuslacerda0) [inline src/middlewares/authMiddleware.ts]: ditto

File: .env [hunk 3]
```diff
@@ -0,0 +1 @@
+SECRET='IOaheihEOHaAOIEHaiohIOEHaioh'
```
Comment (gustavolbs) [inline .env:1]: **nitpick:** Remove this file at https://github.com/TirusTimes/tirus-times-back/pull/2/commits/b98c4a4538fcba3487999b027ea61abe4739074f and add that at `.env.example`

File: src/controllers/AuthControler.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
+import { Request, Response } from "express";
+import { StatusCodes } from "http-status-codes";
+import jwt from 'jsonwebtoken';
+import { prismaClient } from "../database/prismaClient";
+
+export class AuthController {
+    async authenticate(request: Request, response: Response){
+        const { username, password} = request.body;
```
Comment (gustavolbs) [inline src/controllers/AuthControler.ts]:
```suggestion
        const { username, password } = request.body;
```

File: src/controllers/AuthControler.ts [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+        const secret = process.env.SECRET!;
+        const user = await prismaClient.user.findFirst({
+            where: {
+                username
+            }
+        })
+
```
Comment (gustavolbs) [inline src/controllers/AuthControler.ts:15]:
```suggestion
```

File: src/controllers/AuthControler.ts [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+        
+        if (!user) {
+            return response.sendStatus(StatusCodes.UNAUTHORIZED)
+          }
+
+        if (password != user?.password) {
+            return response.sendStatus(StatusCodes.UNAUTHORIZED)
+          }
```
Comment (gustavolbs) [inline src/controllers/AuthControler.ts]:
```suggestion
        if (!user) {
            return response.sendStatus(StatusCodes.UNAUTHORIZED)
        }

        if (password != user?.password) {
            return response.sendStatus(StatusCodes.UNAUTHORIZED)
        }
```

File: src/controllers/AuthControler.ts [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+        delete user.password;
+        
+        return response.json({
+            user,
+            token
+        })
+    }
+
+}
```
Comment (gustavolbs) [inline src/controllers/AuthControler.ts:34]: Please, add newline at EOF

File: src/controllers/AuthControler.ts [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+        const token = jwt.sign({id: user?.id}, secret, { expiresIn: '1d'})
+
+        delete user.password;
+        
+        return response.json({
+            user,
+            token
+        })
+    }
+
```
Comment (gustavolbs) [inline src/controllers/AuthControler.ts:33]:
```suggestion
```

File: src/helpers/schemas.ts [hunk 9]
```diff
@@ -10,5 +10,16 @@ const schemaCreate = Yup.object().shape({
...
-export { schemaCreate }
+const schemaUpdate = Yup.object().shape({
+    username: Yup.string(),
+    firstname: Yup.string(),
+    lastname: Yup.string(),
+    email: Yup.string(),
+    password: Yup.string(),
+    position: Yup.string(),
+    age: Yup.number().positive().integer(),
+    gender: Yup.string()
```
Comment (gustavolbs) [inline src/helpers/schemas.ts:21]: **question:** Does this check works for empty string?

File: src/middlewares/authMiddleware.ts [hunk 10]
```diff
@@ -0,0 +1,22 @@
+import { NextFunction, Request, Response } from "express";
+import { StatusCodes } from "http-status-codes";
+import jwt from 'jsonwebtoken';
+
+export default function authMiddleware(request: Request,response: Response, next: NextFunction){
```
Comment (gustavolbs) [inline src/middlewares/authMiddleware.ts:5]:
```suggestion
export default function authMiddleware(request: Request, response: Response, next: NextFunction){
```

File: src/middlewares/authMiddleware.ts [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+    const { authorization } = request.headers;
+
+    if(!authorization) {
```
Comment (gustavolbs) [inline src/middlewares/authMiddleware.ts:8]:
```suggestion
    if (!authorization) {
```

File: src/middlewares/authMiddleware.ts [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+    const token = authorization.split(' ')[1];
+    const secret = process.env.SECRET!
+
+    try {
+        const data = jwt.verify(token, secret)
+        return next();
+    } catch {
+        return response.sendStatus(StatusCodes.UNAUTHORIZED)
+    }
+
```
Comment (gustavolbs) [inline src/middlewares/authMiddleware.ts]:
```suggestion
```

File: src/middlewares/authMiddleware.ts [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+}
```
Comment (gustavolbs) [inline src/middlewares/authMiddleware.ts]: Please, add newline at EOF

File: src/routes/user.routes.ts [hunk 14]
```diff
@@ -1,13 +1,15 @@
 import { Router } from "express";
 import { userController } from "../controllers/userController";
+import authMiddleware from "../middlewares/authMiddleware";
 const userRoutes = Router();
```
Comment (gustavolbs) [inline src/routes/user.routes.ts:7]:
```suggestion
import authMiddleware from "../middlewares/authMiddleware";

const userRoutes = Router();
```

File: src/services/userService.ts [hunk 15]
```diff
@@ -74,9 +78,14 @@ class userService {
       position,
       age,
       gender,
-    }: IUser
-  ) {
+    } = newData;
+
```
Comment (gustavolbs) [inline src/services/userService.ts]:
```suggestion

```

File: tsconfig.json [hunk 16]
```diff
@@ -4,6 +4,7 @@
       "outDir": "dist",
       "strict": true,
       "lib": ["esnext"],
-      "esModuleInterop": true
+      "esModuleInterop": true,
+      "typeRoots": ["node_modules/yup"]
     }
   }
```
Comment (gustavolbs) [inline tsconfig.json]: Please add newline at EOF

**Review by gustavolbs — CHANGES_REQUESTED** [review]
**Praise:** nice job here!
We just need to do some adjustments

**Pull request merged**